flang/include/flang/Lower/Support/BoxValue.h

```diff
@@ -353,8 +416,9 @@ bool isArray(const ExtendedValue &exv);
 /// indices if it is an array entity.
 class ExtendedValue : public details::matcher<ExtendedValue> {
 public:
-  using VT = std::variant<UnboxedValue, CharBoxValue, ArrayBoxValue,
-                          CharArrayBoxValue, BoxValue, ProcBoxValue>;
+  using VT =
+      std::variant<UnboxedValue, CharBoxValue, ArrayBoxValue, CharArrayBoxValue,
+                   BoxValue, ProcBoxValue, IrBoxValue>;
```

Comment:
>Should MutableBoxValue be in the list?

I think they should, I will probably do this in a separate patch. I refrained doing so because they are few places where it matters that something is allocatable/pointer in expressions, and we mainly want to work with the ExtendedValue resulting from reading the properties of the `MutableBoxValue` symbol. But for allocatable/pointer derived type, that won't completely fly, so I need to add it.

> Since it isn't, it makes me wonder if there is a clear semantic distinction between it and the IrBoxValue class.

They are two main differences:
- `MutableBoxValue` are not always tracked with a `fir.ref<fir.box<T>>` (they always have one in case we need to pass them, but it is not always used for local allocatables, that are tracked with a set of local variable). `IrBoxValue` are meant to be `fir.box<T>`, it makes no sense to track them as a set of local variables (we have the other box value categories for that).
- `MutableBoxValue` associated descriptor is a `fir.ref<fir.box>`, its content may change, `IrBoxValue` descriptor is a `fir.box`, it cannot change.

> any argument may have some of its parameter values overridden in the callee

`overridden` is not the word I would chose here. For an `IrBoxValue`, it's illegal for to specify explicit length parameters that would differ from the ones in the `fir.box`. Explicit length parameters are more a sort of contract/promise that the programmer is making to the compiler (e.g. it may help deducing that length parameters are the same between two entities at compile time, even if we do not know what the value is until runtime).

For `MutableBoxValue`, either the `fir.box` is a dummy that is allocated, in which case it is just like the `IrBoxValue` above, either it is not allocated, and then it actually makes a semantic difference to have explicit length parameters or not. The length parameters are simply not defined yet in the fir.box, so the explicit ones are actual definition we need to keep track of (no definition might be given for them in an allocate-stmt/one can inquire about them before the allocation).

> IrBoxValue seems to factor out the optional property (which is a disassociated reference), but how different is that than a disassociated POINTER argument in terms of its assumed state?

It is different. Reading the descriptor of a disassociated pointer is OK (the resulting values, but for the address field, are undefined, but it is OK to read them). Reading the descriptor from an absent optional is undefined behavior/will lead to segfault (the descriptor of absent optional has no storage).

So I really think POINTER/ALLOCATABLE deserve their own extended value class.

Reply:
> overridden is not the word I would chose here.

It's clearly possible to generally override properties though. We can pass an array[1:100] to a subroutine that maps the indices to an array[-10:-1].

> So I really think POINTER/ALLOCATABLE deserve their own extended value class.

I'm confused. They do: MutableBoxValue, right? But that doesn't fully explain why an IrBoxValue.

Reply:
Sorry, I thought your point was that `IrBoxValue` and `MutableBoxValue` could be the same class, now I think I get that you meant `IrBoxValue` was redundant given `BoxValue`. That point I get, see my other answer above.